src/Symfony/Component/DependencyInjection/ServiceLocator.php

```diff
@@ -142,4 +143,24 @@ private function formatAlternatives(array $alternatives = null, $separator = 'an
 
         return sprintf($format, $alternatives ? implode('", "', $alternatives) : $last, $alternatives ? sprintf(' %s "%s"', $separator, $last) : '');
     }
+
+    public function getProvidedServices(): array
+    {
+        if (null === $this->providedTypes) {
+            $this->providedTypes = array();
+
+            foreach ($this->factories as $name => $factory) {
+                $type = (new \ReflectionFunction($factory))->getReturnType();
+
+                $this->providedTypes[$name] = $type ? $this->determineTypeString($type) : '?';
```

Comment:
let's inline the method, this is fine:
`$this->providedTypes[$name] = $type ? ($type->allowsNull() ? '?' : '').$type->getName() : '?';`